src/lib/protocols/ssh.c

```diff
@@ -57,43 +62,95 @@
 /* #define SSH_DEBUG 1 */
 
 static void ndpi_search_ssh_tcp(struct ndpi_detection_module_struct *ndpi_struct, struct ndpi_flow_struct *flow);
-  
+
 /* ************************************************************************ */
 
-static void ssh_analyse_signature_version(struct ndpi_detection_module_struct *ndpi_struct,
+static void ssh_analyze_signature_version(struct ndpi_detection_module_struct *ndpi_struct,
 					  struct ndpi_flow_struct *flow,
 					  char *str_to_check,
 					  u_int8_t is_client_signature) {
 
+  if (str_to_check == NULL) return;
+
+  int major = 0;
+  int minor = 0;
+  int patch = 0;
+  int obsolete_ssh_version = 0;
+
+  if (sscanf(str_to_check, "SSH-%*f-OpenSSH_%d.%d.%d", &major, &minor, &patch) != 3) 
+    return;
+
+  if ((major || minor || patch) == 0) return;
+
+  /* checking if is an old version */ 
+  if (major < MAJOR_CUTOFF) obsolete_ssh_version = 1;
+
+  else if (major == MAJOR_CUTOFF) {   
+    if (minor < MINOR_CUTOFF) obsolete_ssh_version = 1;
+    
+    else if (minor == MINOR_CUTOFF)
+      if (patch < PATCH_CUTOFF) obsolete_ssh_version = 1;
+  }
   
-  /*
-  if(obsolete_ssh_version)
-    NDPI_SET_BIT(flow->risk, is_client_signature ? NDPI_SSH_OBSOLETE_CLIENT_SIGNATURE : NDPI_SSH_OBSOLETE_SERVER_SIGNATURE);
-  */
+  if (obsolete_ssh_version) {
+      #ifdef SSH_DEBUG
+        printf("[SSH] [SSH Version: %d.%d.%d]\n", major, minor, patch);
+      #endif
+
+      NDPI_SET_BIT(flow->risk, (is_client_signature ? NDPI_SSH_OBSOLETE_CLIENT_VERSION_OR_CIPHER : NDPI_SSH_OBSOLETE_SERVER_VERSION_OR_CIPHER));
+  }
 }
   
 /* ************************************************************************ */
 
 static void ssh_analyse_cipher(struct ndpi_detection_module_struct *ndpi_struct,
 			       struct ndpi_flow_struct *flow,
-			       char *cipher, u_int cipher_len,
+			       char *ciphers, u_int cipher_len,
 			       u_int8_t is_client_signature) {
-  /*
-    List of obsolete ciphers can be found at
-    https://www.linuxminion.com/deprecated-ssh-cryptographic-settings/
-  */
-#ifdef SSH_DEBUG
-  u_int i;
-  
-  printf("[%s] ", is_client_signature ? "CLIENT" : "SERVER");
 
-  for(i=0; i<cipher_len; i++)
-    printf("%c", cipher[i]);
+  char *rem;
+  char *cipher;
+  int found_obsolete_cipher = 0;
 
-  printf("\n");
-#endif
+  char *obsolete_ciphers[6] = {
```

Comment:
Do not put a size (6) but add a NULL at the end of this data structure.Also please use "const char*..."